ext/backtracie_native_extension/c_test_helpers.c

```diff
@@ -0,0 +1,39 @@
+#include "backtracie_private.h"
+#include "public/backtracie.h"
+
+#include <ruby.h>
+
+static VALUE backtracie_backtrace_from_thread(VALUE self);
+static VALUE backtracie_backtrace_from_thread_cthread(void *ctx);
+static VALUE stdlib_backtrace_from_thread(VALUE self);
+static VALUE stdlib_backtrace_from_thread_cthread(void *ctx);
+
+void backtracie_init_c_test_helpers(VALUE backtracie_module) {
+  VALUE test_helpers_mod =
+      rb_define_module_under(backtracie_module, "TestHelpers");
+  rb_define_singleton_method(test_helpers_mod,
+                             "backtracie_backtrace_from_thread",
+                             backtracie_backtrace_from_thread, 0);
+  rb_define_singleton_method(test_helpers_mod, "stdlib_backtrace_from_thread",
+                             stdlib_backtrace_from_thread, 0);
+}
+
+static VALUE backtracie_backtrace_from_thread(VALUE self) {
+  VALUE th = rb_thread_create(backtracie_backtrace_from_thread_cthread, NULL);
+  return rb_funcall(th, rb_intern("value"), 0);
+}
+
+static VALUE backtracie_backtrace_from_thread_cthread(void *ctx) {
+  VALUE backtracie_mod = rb_const_get(rb_cObject, rb_intern("Backtracie"));
+  return rb_funcall(backtracie_mod, rb_intern("backtrace_locations"), 1,
+                    rb_thread_current());
+}
```

Comment:
Actually, now that I look at this I think there's even one more corner case: doing something like `rb_thread_create(backtracie_sleep, ...)` and then having `backtrace_sleep { release_gvl; wait_for_condition_variable; }`.

That way we could test a thread with an entirely empty stack (from a different thread). Right now because we're calling Ruby-level APIs from the thread, there's still "something" that ends up on the stack.

Reply:
Reviewing the comments above I realized I had mentioned a similar situation that happens with the ffi gem. I half-wonder if we'll crash in this case, but I'm happy to merge this PR and find out separately :)

Reply:
Let me find out if it crashes 😂  my money is on no.

Reply:
Looks like i owe you a beer or something in japan.

```
Failures:

  1) Backtracie when sampling a thread from rb_create_thread, with no frames at all returns an empty array
     Failure/Error: let(:backtracie_backtrace) { Backtracie::TestHelpers.backtracie_backtrace_from_empty_thread }

     ArgumentError:
       integer overflow: 18446744073709551615 * 40 > 18446744073709551615
     # ./spec/unit/backtracie_spec.rb:705:in `backtrace_locations'
     # ./spec/unit/backtracie_spec.rb:705:in `backtracie_backtrace_from_empty_thread'
     # ./spec/unit/backtracie_spec.rb:705:in `block (3 levels) in <top (required)>'
     # ./spec/unit/backtracie_spec.rb:708:in `block (3 levels) in <top (required)>'
```

Reply:
And this is why:

```
Process 63062 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = step over
    frame #0: 0x0000000100ff22f4 backtracie_native_extension.bundle`backtracie_frame_count_for_thread [inlined] backtracie_frame_count_for_execution_context(ec=<unavailable>) at backtracie_frames.c:262:7
   259 	  // +2 because of the two dummy frames at the bottom of the stack.
   260 	  const rb_control_frame_t *start_cfp = RUBY_VM_END_CONTROL_FRAME(ec) - 2;
   261
-> 262 	  if (start_cfp < last_cfp) {
   263 	    return 0;
   264 	  } else {
   265 	    return (int)(start_cfp - last_cfp + 1);
Target 0: (ruby) stopped.
(lldb) print start_cfp
(const rb_control_frame_t *) $1 = 0xffffffffffffff80
(lldb) print last_cfp
(const rb_control_frame_t *) $2 = NULL
```

those CFP pointers are hot garbage in the empty-stack case it seems. Should be easy to detect this and return 0 for the frame count in any case.

Reply:
Hah never trust native code not to do something wonky :)